projects/cli-files/go-cat/cmd/root.go

```diff
@@ -0,0 +1,66 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+)
+
+func Execute() {
+	args := os.Args[1:]
+
+	for _, arg := range args {
+		err := checkArgs(arg)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return
+		}
+		fileLines, err := readFileLines(arg)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		for _, fileLine := range fileLines {
+			fmt.Print(fileLine)
+		}
+	}
+}
+
+func checkArgs(arg string) error {
+	if arg == "" {
+		return fmt.Errorf("go-cat: no file specified")
+	}
+	fileInfo, err := os.Stat(arg)
+	if err != nil {
+		return fmt.Errorf("go-cat: '%s': no such file or directory", arg)
+	}
+	if fileInfo.IsDir() {
+		return fmt.Errorf("go-cat: '%s': is a directory", arg)
+	}
+	return nil
+}
+
+func readFileLines(filePath string) ([]string, error) {
```

Comment:
This function seems like quite a complicated way of reading a file - can you talk through why you're going about all of these steps?

Reply:
As you mentioned in the first PR , that `os.ReadFile(path)` read the whole file into memory. I search and found few ways to read the file bit by bit and printing the output straight away using a buffer. I've used `buffio` package for this. It creates a buffer with a specified size from where we read data. Normally, from the examples that I've seen they print the chunk straight away , but I stored them in a slice to be able to test them (before the print).

Reply:
In this case you're actually still reading the whole file into memory because you're accumulating it all in the slice of strings. If you were reading a chunk then immediately printing it and throwing it away, then reading a chunk and immediately printing it and throwing it away, that could work - maybe you can try to restructure your code so that after each time you read something, you quickly get rid of it, rather than keeping ahold of it?

(There's also _another_ way of doing this which doesn't require you to manually buffer - maybe you can try to work that out too, and we can talk through the trade-offs of the two approaches :))